### PR TITLE
feat: Create network routes

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Create the necessary routes to start the UPF services. This also includes creating a new `UPFNetwork` class responsible for all network related operations. This class uses the `pyroute2` and `python-iptables` python packages for network related operations. 

## Notes
- This project does not have real integration tests at the moment because they require having a machine with 2 network interfaces. This is a problem of its own and will be tackled through [TELCO-1013](https://warthogs.atlassian.net/browse/TELCO-1013)

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

[TELCO-1013]: https://warthogs.atlassian.net/browse/TELCO-1013?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/network.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+from typing import List
+
+from machine import ExecError, Machine
+
+logger = logging.getLogger(__name__)
+
+
+class Network:
+    """Abstract the network configuration for the UPF service."""
+
```
Comment (ghislainbourgeois) [inline src/network.py:18]: I am not sure about the way this class is implemented. I agree with having a class that handles the network configuration, but it seems to be acting at different abstraction layers, from the high level configuration, to the low-level interaction with command outputs.

I think splitting this a bit further would also make the tests a bit easier. I would for example highly consider using a library like `pyroute2` to handle the lower level access to interfaces, IP addresses and routes. Similarly, I would consider a library like `python-iptables` to handle the firewall rules.

If we do not want to bring 3rd party libraries as dependencies, I would suggest at least extracting those things to other separate classes. Maybe something like:

```python
nic = Interface('access')
if not nic.exists():
    logger.error("doh")
ip = nic.ip
```

Smaller objects like that would be easier to test, and the testing the higher level logic of testing the network configuration of the whole UPF would need less setup.

What do you think?

**Review by ghislainbourgeois — COMMENTED** [review]
I did not look at the whole PR yet and mostly focused on the `Network` class and tests. I don´t think anything is particularly wrong with the code, but I think the design could be improved; I wrote some thoughts on that for discussion.

File: src/network.py (hunk 1 above)
Comment (gruyaume) [reply src/network.py]: I went ahead and did that, let me know what you think

File: src/upf_network.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        return True
+
+    def get_ip_address(self) -> str:
+        """Get the IP address of the given network interface."""
+        interfaces = self.network_db.interfaces  # type: ignore[reportAttributeAccessIssue]
+        try:
+            iface_record = interfaces[self.name]
+            ip_addresses = iface_record.ipaddr
+            for ip in ip_addresses:
+                if ip.family == 2:
```
Comment (ghislainbourgeois) [inline src/upf_network.py:50]: We can import the constant from the socket module to clarify this magic number:

```
from socket import AF_INET
```

```suggestion
                if ip.family == AF_INET:
```

File: src/upf_network.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                    return ip.address
+            logger.warning("No IPv4 address found for interface %s", self.name)
+        except KeyError:
+            logger.warning("Interface %s not found in the network database", self.name)
+        return ""
+
+    def get_gateway_ip_address(self) -> str:
+        """Get the gateway IP address of the given network interface."""
+        iface_index = self.get_index()
+        routes = self.ip_route.get_routes(family=2)
```
Comment (ghislainbourgeois) [inline src/upf_network.py:60]:
```suggestion
        routes = self.ip_route.get_routes(family=AF_INET)
```

File: tests/unit/test_upf_network.py [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+
+import iptc
+from pyroute2 import NetlinkError
+from upf_network import IPTablesRule, NetworkInterface, Route, UPFNetwork, UPFNetworkError
+
+
+class MockIPAddr:
+    def __init__(self, ipv4_address: str = "", ipv6_address: str = ""):
+        if ipv4_address:
+            self.family = 2
```
Comment (ghislainbourgeois) [inline tests/unit/test_upf_network.py:16]:
```suggestion
            self.family = AF_INET
```

File: tests/unit/test_upf_network.py [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+            self.address = ipv4_address
+        elif ipv6_address:
+            self.family = 10
```
Comment (ghislainbourgeois) [inline tests/unit/test_upf_network.py:19]:
```suggestion
            self.family = AF_INET6
```

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]
I have suggestions to replace some magic numbers, otherwise LGTM

File: src/upf_network.py (hunk 2 above)
Comment (gruyaume) [reply src/upf_network.py]: Done

File: tests/unit/test_upf_network.py (hunk 4 above)
Comment (gruyaume) [reply tests/unit/test_upf_network.py]: Done

File: tests/unit/test_upf_network.py (hunk 5 above)
Comment (gruyaume) [reply tests/unit/test_upf_network.py]: Done

File: src/upf_network.py (hunk 3 above)
Comment (gruyaume) [reply src/upf_network.py]: Done

**Review by ghislainbourgeois — APPROVED** [review]

**Pull request merged**